curated_transformers/models/with_strided_spans.py

```diff
@@ -51,25 +56,92 @@ def with_strided_spans_forward(
 
     Y_layer, backprop_layer = model.layers[0](spans, is_train=is_train)
 
+    # Suppose we have:
+    #
+    # A B C D E F G H I
+    # <--->   stride
+    # <-----> window
+    #
+    # D is part of both the current and the next window. To
+    # ensure that there is no drift in representations, we
+    # average them in both representations.
+    _apply_to_overlaps(
+        Y_layer, doc_lens, stride=stride, window=window, func=_average_arrays
+    )
+
+    Y_docs = _strided_arrays_to_ragged(
+        model, Y_layer, doc_lens, stride=stride, window=window
+    )
+
     def backprop(dY):
         dY_spans, dY_lengths = _ragged_to_strided_arrays(
             dY, stride=stride, window=window
         )
+
+        # Both overlaps will have dY. However, the proper gradient is 0.5 * dY
+        # since we averaged the overlaps in the forward pass.
+        _apply_to_overlaps(
+            Y_layer,
+            doc_lens,
+            stride=stride,
+            window=window,
+            func=_normalize_gradients,
+        )
+
         dXlf = backprop_layer(dY_spans)
         return _strided_arrays_to_ragged(
             model, dXlf, dY_lengths, stride=stride, window=window
         )
 
-    Y_docs = _strided_arrays_to_ragged(
-        model, Y_layer, doc_lens, stride=stride, window=window
-    )
-
     return Y_docs, backprop
 
 
+def _apply_to_overlaps(
+    Xlf: List[Floats2d],
+    lens: List[Ints1d],
+    *,
+    stride: int,
+    window: int,
+    func: Callable,
+):
+    """Average representations of overlapping windows. This function
+    modifies the arrays in Xlf in-place to avoid"""
+
+    if stride > window:
+        raise ValueError(
+            f"Stride size ({stride}) should less than or equal to window size ({window})"
+        )
```

Comment:
Could we move this check to the entry-point, next to `if window // 2 > stride` or merge them (since we don't expect the value to change after model construction)?

Reply:
Moved to the entry point.